zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -0,0 +1,533 @@
+            // (Since the upgraded format is added to the tip, the database can switch between
+            // lower and higher versions at any block.)
+            //
+            // Keep upgrading until the initial database has been upgraded,
+            // or this task is cancelled by a shutdown.
+            while upgrade_height <= initial_tip_height
+                && matches!(cancel_receiver.try_recv(), Err(mpsc::TryRecvError::Empty))
+            {
+                // TODO: Do one format upgrade step here
+                //
+                // This fake step just shows how to access the database.
+                let _replace_me_ = upgrade_db.tip();
+
+                upgrade_height = (upgrade_height + 1).expect("task exits before maximum height");
+            }
+        }
+
+        // At the end of each format upgrade, the database is marked as upgraded to that version.
+        // Upgrades can be run more than once if Zebra is restarted, so this is just a performance
+        // optimisation.
+        info!(
+            ?initial_tip_height,
+            ?newer_running_version,
+            ?older_disk_version,
+            "marking database as upgraded"
+        );
+        Self::mark_as_upgraded_to(&database_format_add_format_change_task, &config, network);
+
+        // End of example format change.
+
+        // # New Upgrades Usually Go Here
+        //
+        // New code goes above this comment!
+        //
+        // Run the latest format upgrade code after the other upgrades are complete,
+        // then mark the format as upgraded. The code should check `cancel_receiver`
+        // every time it runs its inner update loop.
+        info!(
+            ?initial_tip_height,
+            ?newer_running_version,
+            ?older_disk_version,
+            "database is fully upgraded"
+        );
+    }
+
+    /// Mark a newly created database with the current format version.
+    ///
+    /// This should be called when a newly created database is opened.
+    ///
+    /// # Concurrency
+    ///
+    /// The version must only be updated while RocksDB is holding the database
+    /// directory lock. This prevents multiple Zebra instances corrupting the version
+    /// file.
+    ///
+    /// # Panics
+    ///
+    /// If the format should not have been upgraded, because the database is not newly created.
+    fn mark_as_newly_created(config: &Config, network: Network) {
+        let disk_version = database_format_version_on_disk(config, network)
+            .expect("unable to read database format version file path");
+        let running_version = database_format_version_in_code();
+
+        assert_eq!(
+            disk_version, None,
+            "can't overwrite the format version in an existing database:\n\
+             disk: {disk_version:?}\n\
+             running: {running_version}"
+        );
+
+        write_database_format_version_to_disk(&running_version, config, network)
```

Comment:
Do we want to add the format version to RocksDB and use transactions in case it's interrupted during a migration? 

We could update ~100 blocks at a time and update the disk version to something like `v1.0.0(1131..2210), v1.01(0..1131)` leaving the height range in parens during a migration when there are multiple data formats.

Reply:
That's a good question! I'm not sure if we have enough information to answer it yet. (That's why I haven't written up the state upgrade documentation yet.)

Here are my current thoughts:

We could handle partial migrations by expecting upgrade authors to make their upgrade code work for full or partial migrations. Then that code could also handle downgrades to an earlier state version as well. (A downgrade adds new blocks after the tip in an earlier state format, a partial migration leaves some higher or lower blocks in an earlier state format, depending on the height order it does the upgrade in.)

If upgrades need to track how many blocks have been upgraded in order to be correct, we can add that code later. But earlier versions won't track that info, so they'll have to assume any un-marked blocks are not upgraded. (That's a disadvantage of not building it into the upgrade framework from the start.)

I wanted to have the version outside the database for 2 reasons:
- we already have the major version in the directory path outside the database
- if we need to do an upgrade or fix an issue before opening the database, then we might need to know the version to open it correctly (for example, if we need to change the open options or column families)

Transactions can lock the entire database, blocking other reads and writes. So I'd prefer to keep them as small as possible. We could do each height (or other consistent update) in its own transaction, if needed. But none of our planned upgrades need transactions yet.

Here's how the 3 upgrades we have planned for RPCs would work:
- delete redundant tree states: correct if applied multiple times (don't panic if the deleted item isn't there), correct but lowers performance if some duplicate tree states aren't deleted, no transactions needed (single column family), can be done in any order
- add tree sizes: correct if applied multiple times (skip if the size has already been added), no transactions needed (single column family), might need to be added in reverse height order for lightwalletd to do fast sync correctly
- create subtree index: correct if applied multiple times (just writes the same data), no transactions needed (single column family), needs to be added in reverse height order for lightwalletd to do fast sync correctly

Using reverse height order might be the way to go for all upgrades, because the latest blocks are usually the ones being accessed most frequently.

Reply:
Edit: "add tree sizes" isn't a database format change, it's a new API that uses the existing tree data

Reply:
> We could update ~100 blocks at a time and update the disk version to something like `v1(1131..2210), v2(0..1131)` leaving the height range in parens during a migration when there are multiple data formats.

I was thinking of logging progress in a separate text file or the Zebra logs, but I cut it from scope because the rest of the PR was too big!

Reply:
> I'm not sure if we have enough information to answer it yet.
> ..
> If upgrades need to track how many blocks have been upgraded in order to be correct, we can add that code later.

Okay, I was thinking of a disk format change for note commitment trees, but on second thought it's better to let upgrade authors make their upgrade code work for full or partial migrations.

> Here's how the 3 upgrades we have planned for RPCs could work

It seems easy to figure out where it left off by checking the data in these cases, that may be ideal.

Reply:
> > Here's how the 3 upgrades we have planned for RPCs could work
> 
> It seems easy to figure out where it left off by checking the data in these cases, that may be ideal.

I think the subtree indexes will be `index -> (start_height, end_height)`, so we could look up the earliest index key, find the earliest height, and go from there. The others have height keys.